src/Hazelcast.Net/Clustering/ClusterConnections.cs

```diff
@@ -156,129 +196,81 @@ private async ValueTask RaiseConnectionClosed(MemberConnection connection, bool
+            // and still disconnected - if the cluster is down or shutting down, give up
+            if (_clusterState.ClientState != ClientState.Disconnected) 
             {
-                _onClosedMutex.Release();
+                _logger.LogInformation("Disconnected (shutting down)");
+                return default;
             }
 
-            // otherwise, we have lost the last connection
-
-            // --- thread safety ---
-            // although we have released _onClosedMutex, connectAddresses' queue is
-            // empty and there are no connections left, which means that no other
-            // connection can be added/removed at that point
-
-            if (!_clusterState.IsActive || _disposed == 1)
-            {
-                // the cluster is down
-                _logger.LogInformation("Disconnected (down)");
-                await _clusterState.TransitionAsync(ClientState.Shutdown).CfAwait();
-                return;
-            }
-
-            // the cluster is now disconnected, but not down
-            _logger.LogInformation("Disconnected (reconnect mode: {ReconnectMode}, {ReconnectAction})",
+            // the cluster is disconnected, but not down
+            _logger.LogInformation("Disconnected (reconnect mode == {ReconnectMode} => {ReconnectAction})",
                 _clusterState.Options.Networking.ReconnectMode,
                 _clusterState.Options.Networking.ReconnectMode switch
                 {
-                    ReconnectMode.DoNotReconnect => "remain disconnected",
-                    ReconnectMode.ReconnectSync => "not supported -> trying to reconnect asynchronously",
-                    ReconnectMode.ReconnectAsync => "trying to reconnect asynchronously",
-                    _ => "Meh?"
+                    ReconnectMode.DoNotReconnect => "shut down",
+                    ReconnectMode.ReconnectSync => "reconnect synchronously",
+                    ReconnectMode.ReconnectAsync => "reconnect asynchronously",
+                    _ => "meh?"
                 });
 
-
             // what we do next depends on options
             switch (_clusterState.Options.Networking.ReconnectMode)
             {
                 case ReconnectMode.DoNotReconnect:
-                    // DoNotReconnect = the cluster remains unconnected
-                    await _clusterState.TransitionAsync(ClientState.Shutdown).CfAwait();
+                    // DoNotReconnect = the cluster shuts down
+                    _clusterState.RequestShutdown();
                     break;
 
-                case ReconnectMode.ReconnectSync: // TODO: implement ReconnectSync?
-                    // ReconnectSync = the cluster reconnects
-                    // operations are blocked (?) until the cluster is reconnected
-                    // was never supported by the CSharp client
-                    //_clusterState.OperationsShouldWaitIfReconnecting = true;
-
+                case ReconnectMode.ReconnectSync:
```

Comment:
So TODO for `ReconnectSync` is removed but there's no implementation yet if I understand correctly. Does it mean it won't be implemented or TODO was removed accidentally?

Reply:
Not accidental. There is nothing TODO here, in both cases we need to do the same thing. It's somewhere else that we will need to do something. It's when we run an invocation and we are not connected: in one case we should throw, in the other we should wait.